src/main/java/org/openrewrite/staticanalysis/RemoveUnusedParams.java

```diff
@@ -0,0 +1,194 @@
+
+    @RequiredArgsConstructor
+    private static class RemoveUnusedParametersVisitor extends JavaIsoVisitor<ExecutionContext> {
+        private final Accumulator acc;
+
+        @Override
+        public J.MethodDeclaration visitMethodDeclaration(J.MethodDeclaration method, ExecutionContext ctx) {
+            J.MethodDeclaration m = super.visitMethodDeclaration(method, ctx);
+            if (shouldPruneParameters(m, acc)) {
+                List<Statement> prunedParams = filterUnusedParameters(m, collectUsedParameters(m));
+                return prunedParams.equals(m.getParameters()) ? m : m.withParameters(prunedParams);
+            }
+            return m;
+        }
+
+        private boolean shouldPruneParameters(J.MethodDeclaration m, Accumulator acc) {
+            if (m.getBody() == null ||
+                    m.getMethodType() == null ||
+                    m.hasModifier(J.Modifier.Type.Native) ||
+                    !m.getLeadingAnnotations().isEmpty()) {
+                return false;
+            }
+            String signature = MethodMatcher.methodPattern(m.getMethodType());
+            return !acc.overrideSignatures.contains(signature);
+        }
+
+        private Set<String> collectUsedParameters(J.MethodDeclaration m) {
+            Deque<Set<String>> shadowStack = new ArrayDeque<>();
+            return new JavaIsoVisitor<Set<String>>() {
+                @Override
+                public J.Block visitBlock(J.Block block, Set<String> u) {
+                    shadowStack.push(new HashSet<>());
+                    try {
+                        return super.visitBlock(block, u);
+                    } finally {
+                        shadowStack.pop();
+                    }
+                }
+
+                @Override
+                public J.VariableDeclarations visitVariableDeclarations(J.VariableDeclarations decl, Set<String> u) {
+                    decl.getVariables().forEach(v -> shadowStack.peek().add(v.getSimpleName()));
+                    return super.visitVariableDeclarations(decl, u);
+                }
+
+                @Override
+                public J.Identifier visitIdentifier(J.Identifier id, Set<String> u) {
+                    if (isVisibleParameter(id, m, shadowStack)) {
+                        u.add(id.getSimpleName());
+                    }
+                    return id;
+                }
+            }.reduce(m.getBody(), new HashSet<>());
+        }
+
+        private boolean isVisibleParameter(J.Identifier id, J.MethodDeclaration m, Deque<Set<String>> shadowStack) {
+            return !isShadowed(id.getSimpleName(), shadowStack) && isDeclaredAsParameter(id.getSimpleName(), m);
+        }
+
+        private boolean isShadowed(String name, Deque<Set<String>> shadowStack) {
+            for (Set<String> scope : shadowStack) {
+                if (scope.contains(name)) {
+                    return true;
+                }
+            }
+            return false;
+        }
```

Comment:
Instead of looking at shadowed variables I think it's likely safer and easier to use `SemanticallyEqual` by comparing `J.Identifier` instances.